File: src/workflow_terminate/workflow_terminate.erl
-module(workflow_terminate).

% elp:ignore W0012 W0040 E1599
-moduledoc {file, "../../docs/workflow_terminate.md"}.

-export([
    run/0
]).

-spec run() -> temporal_sdk:workflow_result() | no_return().
run() ->
    workflow_eviction:maybe_attach_telemetry(),
    {ok, #{workflow_execution := WE}} = temporal_sdk:start_workflow(
        cluster_1, "default", workflow_terminate_workflow
    ),
    %% Some other work simulated by timer:sleep/1
    timer:sleep(500),
    {ok, #{}} = temporal_sdk:terminate_workflow(cluster_1, WE, [{reason, "test termination"}]),
    {ok, Result} = temporal_sdk:await_workflow(cluster_1, WE),
    Result.

File: lib/workflow_terminate/workflow_terminate.ex
defmodule WorkflowTerminate do
  @external_resource "docs/workflow_terminate.md"
  @moduledoc TemporalSdk.Utils.Code.exdoc!("docs/workflow_terminate.md")

  @spec run() :: :temporal_sdk.workflow_result() | no_return()
  def run() do
    :workflow_eviction.maybe_attach_telemetry()

    {:ok, %{workflow_execution: we}} =
      TemporalSdk.start_workflow(:cluster_1, "default", WorkflowTerminate.Workflow)

    # Some other work simulated by Process.sleep/1
    Process.sleep(500)
    {:ok, %{}} = TemporalSdk.terminate_workflow(:cluster_1, we, reason: "test termination")
    {:ok, result} = TemporalSdk.await_workflow(:cluster_1, we)
    result
  end
end
